src/commit.c

```diff
@@ -302,6 +335,79 @@ int git_commit_amend(
 	return error;
 }
 
+static int insert_commit(git_vector *v, git_repository *repo, git_oid *id)
+{
+	int error;
+	git_commit *commit;
+
+	if ((error = git_commit_lookup(&commit, repo, id)) < 0)
+		return error;
+
+	if ((error = git_vector_insert(v, commit)) < 0) {
+		git_commit_free(commit);
+		return error;
+	}
+
+	return 0;
+}
+
+int git_commit_create_fromstate(
+	git_oid *id,
+	git_repository *repo,
+	const git_signature *author,
+	const git_signature *committer,
+	const char *message_encoding,
+	const char *message)
+{
+	int error;
+	size_t i;
+	git_tree *tree = NULL;
+	git_index *index = NULL;
+	git_oid current_id, tree_id;
+	git_commit *commit;
+	git_vector commits = GIT_VECTOR_INIT;
+	git_repository_state_t state;
+
+	state = git_repository_state(repo);
+
+	if ((error = git_reference_name_to_id(&current_id, repo, GIT_HEAD_FILE)) < 0)
+		return error;
+
+	if ((error = insert_commit(&commits, repo, &current_id)) < 0)
+		goto cleanup;
+
+	/* TODO: how do we handle octopus merges? */
+	if (state == GIT_REPOSITORY_STATE_MERGE) {
+		if ((error = git_reference_name_to_id(&current_id, repo, GIT_MERGE_HEAD_FILE)) < 0)
```

Comment:
In the octopus merge case, `MERGE_HEAD` will contain the id of each head being merged one per line.  I suspect that `reference_name_to_id` will fail on this as the file will have what it believes to be trailing garbage.